editor/utils/block-list.js

```diff
@@ -32,34 +41,94 @@ const INNER_BLOCK_LIST_CACHE = {};
  */
 export function createInnerBlockList( uid, renderBlockMenu, showContextualToolbar ) {
 	if ( ! INNER_BLOCK_LIST_CACHE[ uid ] ) {
-		INNER_BLOCK_LIST_CACHE[ uid ] = [
-			// The component class:
-			class extends Component {
-				componentWillMount() {
-					INNER_BLOCK_LIST_CACHE[ uid ][ 1 ]++;
+		const InnerBlockListComponent = class extends Component {
+			constructor() {
+				super( ...arguments );
+				this.updateNestedSettings = this.updateNestedSettings.bind( this );
+				this.insertTemplateBlocks = this.insertTemplateBlocks.bind( this );
+			}
+
+			insertTemplateBlocks( template ) {
+				const { block, insertBlocks } = this.props;
+				if ( template && ! block.innerBlocks.length ) {
+					// synchronizeBlocksWithTemplate( [], template ) parses the template structure,
+					// and returns/creates the necessary blocks to represent it.
+					insertBlocks( synchronizeBlocksWithTemplate( [], template ) );
 				}
+			}
 
-				componentWillUnmount() {
-					// If, after decrementing the tracking count, there are no
-					// remaining instances of the component, remove from cache.
-					if ( ! INNER_BLOCK_LIST_CACHE[ uid ][ 1 ]-- ) {
-						delete INNER_BLOCK_LIST_CACHE[ uid ];
-					}
+			updateNestedSettings( newSettings ) {
+				if ( ! isEqual( this.props.blockListSettings, newSettings ) ) {
+					this.props.updateNestedSettings( newSettings );
 				}
+			}
 
-				render() {
-					return (
-						<BlockList
-							rootUID={ uid }
-							renderBlockMenu={ renderBlockMenu }
-							showContextualToolbar={ showContextualToolbar }
-							{ ...this.props } />
-					);
+			componentWillMount() {
```

Comment:
[`componentWillMount` will be marked as unsafe in React 17.](https://reactjs.org/docs/react-component.html#unsafe_componentwillmount) Let's future-proof this code by using `componentDidMount` instead.